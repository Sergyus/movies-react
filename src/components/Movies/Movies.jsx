import React, { useState } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames/bind';
import styles from './Movies.scss';
import Grid from './components/Grid';
import Table from './components/Table';
import Spinner from '../Spinner';
import TrailerModal from './components/TrailerModal';

const cx = cn.bind(styles);

const Movies = ({
  isLoading, movies, layout, videoKey, page,
}) => {
  const [isModalOpen, setModalOpen] = useState(false);
  const isMoviesExist = movies.length !== 0;
  let renderContent;

  if (!isMoviesExist) {
    renderContent = <h2>Sorry, no movies found.</h2>;
  } else if (layout === 'grid') {
    renderContent = <Grid toggleModal={setModalOpen} movies={movies} />;
  } else if (layout === 'table') {
    renderContent = <Table toggleModal={setModalOpen} movies={movies} />;
  }

  return (
    <>
      {
        <main className={cx({ container: true, gridMovieList: (layout === 'grid' && isMoviesExist), tableMovieList: (layout === 'table' && isMoviesExist) })}>
          { renderContent }
          {page > 1 ? <a className={styles.toTopBtn} type="button" href="#menu">&#x2B06;</a> : null}
        </main>
      }
      {isModalOpen ? (
        <TrailerModal
          isLoading={isLoading}
          toggleModal={setModalOpen}
          videoKey={videoKey}
        />
      ) : null}
      { isLoading ? <Spinner /> : null }
    </>
  );
};

Movies.propTypes = {
  isLoading: PropTypes.bool,
  movies: PropTypes.arrayOf(PropTypes.object),
  layout: PropTypes.string,
  videoKey: PropTypes.string,
  page: PropTypes.number,
};

Movies.defaultProps = {
  movies: [],
  isLoading: false,
  layout: 'Grid',
  videoKey: '',
  page: 1,
};

export default Movies;
