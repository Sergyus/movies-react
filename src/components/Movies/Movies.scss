@import "../../styles/variables.scss";

.gridMovieList {
  background-size: cover;
  display: grid;
  grid-template-columns: repeat(auto-fill, 250px);
  grid-gap: 2rem;
  justify-content: space-between;
}

.container {
  margin-left: 5%;
  margin-right: 5%;
  margin-bottom: 50px;
}

.toTopBtn {
  position: fixed;
  color: $blue-color;
  text-decoration: none;
  right: 1vw;
  top: 90vh;
  font-size: 3.5rem;
  transition: font-size 0.2s;
}

.toTopBtn:hover {
  font-size: 4rem;
}

@media (max-width: $desktop-screen-size) {
  .gridMovieList {
    justify-content: space-around;
    margin-bottom: 20px;
  }
}

@media (max-width: $mobile-screen-size) {
  .toTopBtn {
    right: 5vw;
  }
}
