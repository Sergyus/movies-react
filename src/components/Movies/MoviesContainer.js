import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import Movies from './Movies';
import {
  fetchTopRatedMovies, fetchLatestMovies, fetchUpcomingMovies,
  fetchSearchMovies, fetchMoviesByGenre, filterSelector,
  currentPageSelector, isLoadingSelector, currentGenreSelector,
  currentSearchQuerySelector, hasMorePagesSelector, videoKeySelector,
  moviesSelector,
} from '../../modules/app';
import { layoutSelector } from '../../modules/ui';
import { useFetchOnScroll } from '../../utils';

const MovieContainer = ({
  isLoading, filter, currentPage, currentGenre,
  currentSearchQuery, movies, layout, getTopRatedMovies,
  getLatestMovies, getUpcomingMovies, getSearchMovies,
  getMoviesByGenre, videoKey, hasMorePages,
}) => {
  useFetchOnScroll(
    isLoading, hasMorePages, filter, currentPage,
    getLatestMovies, getTopRatedMovies, getUpcomingMovies,
    getMoviesByGenre, currentGenre, getSearchMovies, currentSearchQuery,
  );
  return (
    <Movies
      page={currentPage}
      videoKey={videoKey}
      isLoading={isLoading}
      movies={movies}
      layout={layout}
    />
  );
};

const mapStateToProps = state => ({
  movies: moviesSelector(state),
  layout: layoutSelector(state),
  currentPage: currentPageSelector(state),
  isLoading: isLoadingSelector(state),
  filter: filterSelector(state),
  currentGenre: currentGenreSelector(state),
  currentSearchQuery: currentSearchQuerySelector(state),
  hasMorePages: hasMorePagesSelector(state),
  videoKey: videoKeySelector(state),
});

const actions = {
  getTopRatedMovies: fetchTopRatedMovies,
  getLatestMovies: fetchLatestMovies,
  getUpcomingMovies: fetchUpcomingMovies,
  getSearchMovies: fetchSearchMovies,
  getMoviesByGenre: fetchMoviesByGenre,
};

MovieContainer.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  layout: PropTypes.string.isRequired,
  currentPage: PropTypes.number.isRequired,
  filter: PropTypes.string.isRequired,
  currentGenre: PropTypes.string,
  currentSearchQuery: PropTypes.string,
  videoKey: PropTypes.string,
  hasMorePages: PropTypes.bool.isRequired,
  movies: PropTypes.arrayOf(PropTypes.object),
  getTopRatedMovies: PropTypes.func.isRequired,
  getLatestMovies: PropTypes.func.isRequired,
  getUpcomingMovies: PropTypes.func.isRequired,
  getSearchMovies: PropTypes.func.isRequired,
  getMoviesByGenre: PropTypes.func.isRequired,
};

MovieContainer.defaultProps = {
  currentGenre: '',
  currentSearchQuery: '',
  videoKey: '',
  movies: [],
};

export default connect(mapStateToProps, actions)(MovieContainer);
