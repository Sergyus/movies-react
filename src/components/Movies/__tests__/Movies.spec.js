import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Movies from '../Movies';

describe('Movies component renders correctly', () => {
  it('Movies grid renders correctly', () => {
    const renderer = new ShallowRenderer();
    const result = renderer.render(
      <Movies
        page={2}
        movies={[{ title: 'first movie' }, { title: 'second movie' }]}
        layout="grid"
      />,
    );
    expect(result).toMatchSnapshot();
  });
});
