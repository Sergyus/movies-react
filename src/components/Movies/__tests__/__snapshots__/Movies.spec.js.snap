// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movies component renders correctly Movies grid renders correctly 1`] = `
<React.Fragment>
  <main
    className="container gridMovieList"
  >
    <Grid
      movies={
        Array [
          Object {
            "title": "first movie",
          },
          Object {
            "title": "second movie",
          },
        ]
      }
      toggleModal={[Function]}
    />
    <a
      className="toTopBtn"
      href="#menu"
      type="button"
    >
      â¬†
    </a>
  </main>
</React.Fragment>
`;
